### ISSUE TITLE
Dependency Injection class rewrites

### ISSUE THREAD

**Issue by luckyraul** [vimeo/psalm#1226]
How to handle dependency injection with replacing interface with class ?

For an example in Magento I should replace \Magento\Core\Model\UrlInterface with \Magento\Core\Model\Url
```
<preference for="Magento\Core\Model\UrlInterface" type="Magento\Core\Model\Url" />
```

**Comment by muglug** [comment]
How is this used in the codebase?

**Comment by luckyraul** [comment]
You have the xml file with a mapping and Magento core replaces interfaces or classes with mapped ones.

**Comment by muglug** [comment]
Ok, I had a look at the docs - I bet it uses `class_alias`. So for Psalm to understand this it'd need to load the same autoloader that Magento uses. If you have PHPUnit tests you'd have to do something similar.

You can add `autoloader="path_to_file.php"` in the Psalm config.

**Comment by luckyraul** [comment]
As far as I'm concerned for Unit testing Magento uses special ObjectManager that can retrieve actual classes from the class name or interface.

Especially in modules where you don't have full magento stack.

**Comment by muglug** [comment]
I don't really know how Magento works, but you might want to have a look at [this gist](https://gist.github.com/Vinai/9734553) and http://techjeffyu.com/blog/integrating-phpunit-test-in-the-magento-system for guidance - the bootstrap files they're creating are for PHPUnit's benefit, but Psalm uses the same basic idea.

**Comment by weirdan** [comment]
Do you have a specific code example where this would be needed? In a typical (I'm not familiar with Magento though) DIC-managed system consumers should not rely on getting the specific implementation where they declare dependency on an interface. This goes against the whole idea of IoC.

**Comment by weirdan** [comment]
With that said, it's probably possible to create a plugin that would tell Psalm: where `ISomething` is expected, `ISomething&TheSomething` would be supplied, based on some external config.

**Comment by luckyraul** [comment]
I agree with your comment about IoC. But in this test, I do not need whole Magento to be installed with compiled DI. 

In long term, we probably can use a full stack to test one small extension. 

I think currently ```ISomething&TheSomething``` will be enough to test with external DI config. 
I can not find in the plugin documentation that event should I use in a plugin and what should be returned. 
Can you suggest which event I should try?

**Comment by luckyraul** [comment]
https://travis-ci.org/mygento/module-base/jobs/482652383#L1026

**Comment by luckyraul** [comment]
Mygento\Base\Api\Data\EventInterface -> Mygento\Base\Model\Event

Mygento\Base\Model\Event extends Magento\Framework\Model\AbstractModel implements Mygento\Base\Api\Data\EventInterface

**Comment by luckyraul** [comment]
return $this - good idea, thanks. 

What about https://travis-ci.org/mygento/module-base/jobs/482652383#L1089 ?

**Comment by muglug** [comment]
@luckyraul see above to @weirdan's comment - that failure is because it expects `AbstractModel`, and `AbstractModel` doesn't implement `EventInterface` (because that's an interface you created yourself). @weirdan's suggested fix - `/** @psalm-param EventInterface&AbstractModel $event */` - should work there.

**Comment by luckyraul** [comment]
Do Not helped https://travis-ci.org/mygento/module-base/jobs/485167372#L972

**Comment by luckyraul** [comment]
```
/**
     * @param \Mygento\Base\Api\Data\EventInterface $entity
     * @throws \Magento\Framework\Exception\CouldNotSaveException
     * @return \Mygento\Base\Api\Data\EventInterface
     * @psalm-param EventInterface&AbstractModel $entity
     */
    public function save(\Mygento\Base\Api\Data\EventInterface $entity)
    {
        try {
            /** @psalm-param EventInterface&AbstractModel $entity */
            $this->resource->save($entity);
        } catch (\Exception $exception) {
            throw new \Magento\Framework\Exception\CouldNotSaveException(
                __($exception->getMessage())
            );
        }
        return $entity;
    }
```

**Comment by luckyraul** [comment]
@muglug @weirdan Any thoughts?

**Comment by luckyraul** [comment]
@muglug
